### DISCUSSION TITLE
Does ONNX Cater for normalization parameters of the model?

### DISCUSSION THREAD

**Discussion by CJCombrink** [onnx/onnx#2928] [Q&A]
# Does ONNX Cater for normalization parameters of the model?

### Question
Saving a model to ONNX format is one part of the problem. When using a model, it is critically important to feed the model with the correct data, correctly normalized (preprocessing information). 
How does the ONNX specification cater for this, to:
1. Make sure the consumer of the model knows what data goes where.
2. Make sure the consumer of the model knows how the data must be normalized (also if the data must be normalized).

### Further information
If I look at the [MobileNet ](https://github.com/onnx/models/tree/master/vision/classification/mobilenet#preprocessing) model in the model zoo, I see that the README.md document states under the section "Preprocessing":
> The images have to be loaded in to a range of [0, 1] and then normalized using mean = [0.485, 0.456, 0.406] and std = [0.229, 0.224, 0.225]. The transformation should preferrably happen at preprocessing. Check imagenet_preprocess.py for code.

Thus the question, is there a provision in the specification for that information, without one having to read through the code or the README.md. Any one should see the issue with having to read it from there and then update the using code, especially if not familiar with the framework where the model originated from.

The other issue is the data ordering. For example a time-series model predicting a temperature. If the model has a input shape of [N, 1, 8], is there a specific place where this can be specified to state the order (eg. Temperature, Air pressure, .etc.) or is there a best practice approach only (a Readme file, examine the training code, etc.)/ 

### Notes
I am not a data scientist, but need to consume a model and data and produce a prediction, thus having access to this information is essential.

**Comment by skottmckay** [comment]
https://github.com/lutzroeder/netron is a good way to view a model and see the model inputs and their shapes.

There's no provision in the onnx spec for this sort of information. It's up to the model author to determine how the input should be structured and how they're going to communicate that to the user (e.g. via a readme or example code or some other way). If it was part of the spec the info would be embedded in the model, which would arguably be less discoverable and useful.
